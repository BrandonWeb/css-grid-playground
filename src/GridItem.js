import React from 'react';
import styled from 'styled-components';
import TextArea from './TextArea';
import { isValidCss } from './utils';

const Item = styled.div`
  position: relative;
  transition: all 300ms ease-in-out;
  padding: .5rem;
  > textarea {
    transition: all 300ms ease-in-out;
    transition-delay: 500ms;
    opacity: ${props => (props.autoHide ? 0 : 1)};
  }
  &:hover {
    outline: 1px dashed lightgray;
    > textarea {
      height: 100%;
      width: 100%;
      transition-delay: 0s;
      opacity: 1;
    }
  }
  &:hover > div {
    opacity: 1;
    transition-delay: 0s;
  }
  ${props => props.globalItemStyle}
  ${props => props.itemStyle}
`;

const ItemNumberContainer = styled.div`
  position: absolute;
  display: flex;
  justify-content: center;
  align-items: center;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  pointer-events: none;
  transition: all 300ms ease-in-out;
  transition-delay: 500ms;
  opacity: ${props => (props.autoHide ? 0 : 1)};
`;

const ItemNumber = styled.div`
  font-size: 2em;
  color: white;
  opacity: .5;
`;

class GridItem extends React.Component {
  static propTypes = {
    autoHide: React.PropTypes.bool,
    itemStyle: React.PropTypes.string,
    globalItemStyle: React.PropTypes.string,
    updateItemStyle: React.PropTypes.func,
    itemNumber: React.PropTypes.number,
  };
  static defaultProps = {
    autoHide: true,
    itemStyle: '',
    globalItemStyle: '',
    updateItemStyle: null,
    itemNumber: null,
  };
  updateItemStyle = (value) => {
    if (isValidCss(value)) this.props.updateItemStyle(this.props.itemNumber, value);
  }
  render() {
    return (
      <Item
        autoHide={this.props.autoHide}
        itemStyle={this.props.itemStyle}
        globalItemStyle={this.props.globalItemStyle}
      >
        <TextArea
          defaultValue={this.props.itemStyle}
          onChange={this.updateItemStyle}
        />
        <ItemNumberContainer autoHide={this.props.autoHide}>
          <ItemNumber>{this.props.itemNumber + 1}</ItemNumber>
        </ItemNumberContainer>
      </Item>
    );
  }
}

export default GridItem;
