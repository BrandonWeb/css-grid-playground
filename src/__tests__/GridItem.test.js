import React from 'react';
import renderer from 'react-test-renderer';
import GridItem from '../GridItem';

it('renders correctly as item 1', () => {
  const Component = (
    <GridItem
      key={1}
      itemNumber={0}
      itemStyle={'background: red;'}
      autoHide
      globalItemStyle={'height: 100px;'}
      updateItemStyle={() => {}}
    />
  );
  const tree = renderer.create(Component).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders correctly as item 2', () => {
  const Component = (
    <GridItem
      key={1}
      itemNumber={1}
      itemStyle={'background: red;'}
      autoHide
      globalItemStyle={'height: 100px;'}
      updateItemStyle={() => {}}
    />
  );
  const tree = renderer.create(Component).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders grid item with auto hide', () => {
  const Component = <GridItem key={1} itemNumber={1} autoHide />;
  const tree = renderer.create(Component).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders grid item without auto hide', () => {
  const Component = <GridItem key={1} itemNumber={1} />;
  const tree = renderer.create(Component).toJSON();
  expect(tree).toMatchSnapshot();
});

// it('grid item style update', () => {
//   const component = shallow(
//     <GridItem
//       key={1}
//       itemNumber={1}
//       updateItemStyle={() => 'foo'}
//     />,
//   );
//   component.find('textarea').simulate('click').simulate('keydown', { which: 'a' });
//   expect(component.find('textarea').text()).toEqual('a');
//   // const result = Component.props.updateItemStyle();
//   // expect(result).toBe('foo');
// });
