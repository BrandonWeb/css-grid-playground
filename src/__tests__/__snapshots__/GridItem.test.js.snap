exports[`test renders correctly as item 1 1`] = `
<div
  className="sc-bwzfXH iAgeiX">
  <textarea
    className="sc-bdVaJa hmEYoi"
    onChange={[Function]}
    value="background: red;" />
  <div
    className="sc-htpNat iUgnEl">
    <div
      className="sc-bxivhb fEzLwQ">
      1
    </div>
  </div>
</div>
`;

exports[`test renders correctly as item 2 1`] = `
<div
  className="sc-bwzfXH iAgeiX">
  <textarea
    className="sc-bdVaJa hmEYoi"
    onChange={[Function]}
    value="background: red;" />
  <div
    className="sc-htpNat iUgnEl">
    <div
      className="sc-bxivhb fEzLwQ">
      2
    </div>
  </div>
</div>
`;

exports[`test renders grid item with auto hide 1`] = `
<div
  className="sc-bwzfXH dhqRvj">
  <textarea
    className="sc-bdVaJa hmEYoi"
    onChange={[Function]}
    value="" />
  <div
    className="sc-htpNat iUgnEl">
    <div
      className="sc-bxivhb fEzLwQ">
      2
    </div>
  </div>
</div>
`;

exports[`test renders grid item without auto hide 1`] = `
<div
  className="sc-bwzfXH dhqRvj">
  <textarea
    className="sc-bdVaJa hmEYoi"
    onChange={[Function]}
    value="" />
  <div
    className="sc-htpNat iUgnEl">
    <div
      className="sc-bxivhb fEzLwQ">
      2
    </div>
  </div>
</div>
`;
